include/logging.h

```diff
@@ -88,11 +91,36 @@ void GFX_ShowMsg(char const* format,...) GCC_ATTRIBUTE(__format__(__printf__, 1,
 #endif // C_DEBUG
 
 #ifdef NDEBUG
-// DEBUG_LOG_MSG exists only for messages useful during development, and not to
+// LOG_DEBUG exists only for messages useful during development, and not to
 // be redirected into internal DOSBox debugger for DOS programs (C_DEBUG feature).
-#define DEBUG_LOG_MSG(...)
+#define LOG_DEBUG(...)
 #else
-#define DEBUG_LOG_MSG(...) DLOG_F(INFO, __VA_ARGS__)
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wformat-security"
```

Comment:
I tried using `format_string` from `string_utils.h`, but that's a no-go because of circular dependency include hell or something.

Ultimately, I'm completely fine with shutting up those security warnings here because we're in 100% control over the format strings. That's a generic warnings that is not applicable for this specific use case at all (plus we're not OpenSSL 😅).

If someone has a "better" solution that I can just paste in then I will, but I won't spend more time on this as it works and is fit for purpose I think.

Reply:
Suggest moving `-Wformat-security` to meson.build flags, and avoid adding a new pragma that (mostly) touches all the codebase.

Currently the only `#pragma` used is when we need packed structs to match the DOS-side struct layout (and only for on Microsoft Visual Studio):

**`git grep '^#pragma' src/ *.cpp *.c | grep -v /libs/`**

```c++
src/dos/dos_execute.cpp:#pragma pack(1)
src/dos/dos_execute.cpp:#pragma pack()
src/dos/dos_files.cpp:#pragma pack (1)
src/dos/dos_files.cpp:#pragma pack()
src/dos/dos_mscdex.cpp:	#pragma pack(1)
src/dos/dos_mscdex.cpp:	#pragma pack()
src/dos/dos_tables.cpp:#pragma pack(1)
src/dos/dos_tables.cpp:#pragma pack ()
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/hardware/imfc.cpp:#pragma pack(push, 1)
src/hardware/imfc.cpp:#pragma pack(pop)
src/ints/int10_misc.cpp:#pragma pack(1)
src/ints/int10_misc.cpp:#pragma pack()
src/ints/int10_vesa.cpp:#pragma pack (1)
src/ints/int10_vesa.cpp:#pragma pack()
src/ints/xms.cpp:#pragma pack (1)
src/ints/xms.cpp:#pragma pack ()
```

Reply:
> Suggest moving `-Wformat-security` to meson.build flags, and avoid adding a new pragma that (mostly) touches all the codebase.
> 
> Currently the only `#pragma` used is when we need packed structs to match the DOS-side struct layout (and only for on Microsoft Visual Studio):

But why is that approach preferable or better? This way the warning is localised to the `LOG_DEBUG` and `LOG_TRACE` definitions with laser focus.

Ignoring these warnings at the meson.build level would turn them off wholesale? Are you saying we don't care about these security warnings in general anyway because DOSBox does not have high security requirements? (Which I can agree with.)

In more general, judicious use of pragmas is fine I think, ahen you need them you need them. I wouldn't avoid them as a matter of principle.

Reply:
Btw, there's more in `include`. And even more in `scr/libs`. We're already disabling `-Wformat-security` in `string_utils.h`.

<img width="802" alt="image" src="https://github.com/dosbox-staging/dosbox-staging/assets/698770/7690dd2e-77b6-46ef-939c-809e37f4b4a0">

Reply:
> avoid adding a new pragma that (mostly) touches all the codebase.

To be clear, I don't think that's true. That warnings is only turned off `inside` those debug and trace log statements, effectively. It does not matter that they're included say 20k times in other files; outside of the debut and trace statements, those are completely unaffected.

Reply:
> But why is that approach preferable or better? 

I'm not a fan of pragma because it's a pre-processor directive that "branches" at the compiler level, inside the code itself.  The branch being those compilers that recognize the pragma versus those that don't (I'm just speaking in general).

Given any equivalent alternative that's more universal (and not in the source files), I'd rather go that route. So moving this into meson with other C-flags is one way.

> Ignoring these warnings at the meson.build level would turn them off wholesale? Are you saying we don't care about these security warnings in general anyway

Yup.  We're already knocking them in string_utils.h and now in logging.h: both of these headers get included quite high up to the point where (almost all the time), this flag is going into be in-effect.

Plus, if it was historically already such a problem we had to knock it out from our string_util's dealings, then I think it's give us enough headaches already. Time for it to be tossed entirely.

Microsoft's similar "use snprintf" warnings were in a similar boat and we had to shut them off too. Thankfully with static analyzers plus ASAN builds, we have no problems catching any nasty C-string memory hickups!

> In more general, judicious use of pragmas is fine I think; when you need them you need them. I wouldn't avoid them as a matter of principle.

My principle is to avoid them as strongly as other preprocessor directives: move as much as you can into agnostics standard C++. But yeah: if you /need/ to use 'em, then so be it!

Some teams buy heavily into #pragma-based coding styles.
For example:

```c++
#pragma startup my_function 3
#pragma exit other_function 1
```

Where you can control which function(s) run at startup (before the main function is called) and program exit (just before the program terminates through _exit). 

The numbers even lets you control the order (the numbers)! So these become band-aids for badly managed flow-control. 

Whats worse.. they're only honored by Clang and GCC - so MSVC will break :-)

Reply:
> even more in scr/libs

Yup, that's (in theory) 3rd party code that we're trying to avoid maintaining and owning.  So we just copy in new updates from upstream.. and they're free to use pragmas.

The downside is that sloppy pragma's in their headers can affect our own .cpp when we include them. But thankfully they wrap their pragma's inside even more pre-processor protection layers:

enet.h:

```c++
#if defined(_WIN32)
    #if defined(_MSC_VER) && defined(ENET_IMPLEMENTATION)
        #pragma warning (disable: 4267) // size_t to int conversion
        #pragma warning (disable: 4244) // 64bit to 32bit int
        #pragma warning (disable: 4018) // signed/unsigned mismatch
        #pragma warning (disable: 4146) // unary minus operator applied ```

Those thankful only kick in inside the `ENET_IMPLEMENTATION` chunk, which is their .cpp file. But yeah.. messy stuff. Just fix the code guys!

Reply:
Yeah I'm fine with turning them off wholesale. Just wasn't sure about that, that's all, as these pragmas only affect the log statements.

Yeah, I agree pragmas are not great when overused. But not all projects are dealing with server-grade stuff, plus users doing snprintf based attacks is such a moot point in DOSBox when you can make them run arbitrary code inside the emulator anyway 😄 

Ok, removing the warning 😄

Reply:
Ok, removing the warning 😄

(:rocket: ! blow it away from string_utils.h, too)